atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java

```diff
@@ -62,24 +64,146 @@ public SchemaMutationLock(
+
+                    if (Thread.currentThread().isInterrupted()) {
+                        return null;
+                    }
+
+                    CASResult casResult = writeDdlLockWithCas(client, expected, ourUpdate);
+
+                    if (casResult.isSuccess()) {
+                        heartbeatCount.incrementAndGet();
+                    } else {
+                        handleForcedLockClear(casResult, lockId, heartbeatCount.get());
+                    }
+                    return null;
+                });
+            } catch (TException e) {
+                throw Throwables.throwUncheckedException(e);
+            }
+
+        }
+    }
+
+    private class HeartbeatService {
+        private ScheduledExecutorService heartbeatExecutorService;
+        private AtomicInteger heartbeatCount;
+        private final int heartbeatTimePeriodMillis;
+
+        HeartbeatService() {
+            heartbeatCount = new AtomicInteger(0);
+            heartbeatTimePeriodMillis = configManager.getConfig().heartbeatTimePeriodMillis();
+        }
+
+        void startBeatingForLock(long lockId) {
+            String err = "Can\'t start new heartbeat with an existing heartbeat. Only one heartbeat per lock allowed.";
+            Preconditions.checkState(heartbeatExecutorService == null, err);
+
+            heartbeatExecutorService = Executors.newSingleThreadScheduledExecutor();
+            this.heartbeatCount.set(0);
+            heartbeatExecutorService.scheduleAtFixedRate(new Heartbeat(lockId, heartbeatCount), 0,
+                    heartbeatTimePeriodMillis, TimeUnit.MILLISECONDS);
+        }
+
+        void stopBeating() {
+            String err = "Can\'t stop non existent heartbeat.";
+            Preconditions.checkState(heartbeatExecutorService != null, err);
+            heartbeatExecutorService.shutdown();
+            try {
+                if (!heartbeatExecutorService.awaitTermination(10, TimeUnit.SECONDS)) {
+                    heartbeatExecutorService.shutdownNow();
+                    if (!heartbeatExecutorService.awaitTermination(10, TimeUnit.SECONDS)) {
+                        throw new RuntimeException("Could not kill heartbeat");
+                    }
+                }
+            } catch (InterruptedException e) {
+                heartbeatExecutorService.shutdownNow();
+                Thread.currentThread().interrupt();
+            } finally {
+                heartbeatExecutorService = null;
+            }
+        }
+    }
 
+    void runWithLock(Action action) {
+        if (!supportsCas) {
+            runWithLockWithoutCas(action);
+            return;
+        }
+
+        long lockId = waitForSchemaMutationLock();
+        heartbeatService.startBeatingForLock(lockId);
         try {
             action.execute();
         } catch (Exception e) {
             throw Throwables.throwUncheckedException(e);
         } finally {
+            heartbeatService.stopBeating();
```

Comment:
In that scenario, any other client that's waiting on the lock will be able to grab it, despite `schemaMutationUnlock` not being called (since the heartbeat will die). In fact, we no longer need to unlock our locks, simply killing the heartbeat is sufficient to let others grab the lock.